### PR TITLE
Fix/metric tests

### PR DESCRIPTION
**Pull request by sheridavandenbent**
# Results of merging this

Creates more tests for realised drawdown, and incorporates tests for worst/best trade, and marketchanges. Code by @noudcorten .

# What has been changed?

# Examples

# Actions to be performed before this can be merged

# Security Measures

# Potential Issues / Future considerations

# Assumptions made / Things to add to the wiki
 

# Additional Info

### CODE REVIEW

File: cli/checks/latest_version.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import json
 from typing import Optional
-from urllib.error import HTTPError
+from urllib.error import URLError
```
Comment (sheridavandenbent) [inline cli/checks/latest_version.py:3]: Why is this changed? What does it fix?

Comment (sheridavandenbent) [reply cli/checks/latest_version.py:3]: @noudcorten

File: modules/stats/stats.py [hunk 2]
```diff
@@ -52,13 +52,14 @@ def generate_backtesting_result(self,
                                     market_change: dict) -> TradingStats:
 
         trading_module = self.trading_module
+        coin_res, best_worst_trade = self.generate_coin_results(trading_module.closed_trades, market_change)
+        open_trade_res = generate_open_trades_results(trading_module.open_trades)
```
Comment (geordle) [inline modules/stats/stats.py:56]: I believe arg `closed_trades` passed here is only used in `generate_coin_results` for calling `self.calculate_statistics_per_coin`. Maybe lift that call  here. Saves you returning `best_worst_trade`

File: modules/output/results.py [hunk 3]
```diff
@@ -35,8 +35,8 @@ class MainResults:
     n_trades_with_loss: int
     n_consecutive_losses: int
     max_realised_drawdown: float
-    max_drawdown_single_trade: float
-    max_win_single_trade: float
+    worst_trade: float
```
Comment (geordle) [inline modules/output/results.py:38]:
```suggestion
    worst_trade_profit_percentage: float
```

File: modules/stats/stats.py [hunk 4]
```diff
@@ -72,20 +73,26 @@ def generate_backtesting_result(self,
...
         drawdown_from = datetime.fromtimestamp(max_seen_drawdown['from'] / 1000) \
             if max_seen_drawdown['from'] != 0 else '-'
         drawdown_to = datetime.fromtimestamp(max_seen_drawdown['to'] / 1000) \
             if max_seen_drawdown['to'] != 0 else '-'
         drawdown_at = datetime.fromtimestamp(max_seen_drawdown['at'] / 1000) \
             if max_seen_drawdown['at'] != 0 else '-'
+        best_trade = (best_worst_trade['best_trade'] - 1) * 100 \
+            if best_worst_trade['best_trade'] != -np.inf else 0
+        worst_trade = (best_worst_trade['worst_trade'] - 1) * 100 \
+            if best_worst_trade['worst_trade'] != np.inf else 0
```
Comment (geordle) [inline modules/stats/stats.py:96]: Split `best_worst_trade` into 2 vars

File: modules/stats/stats.py [hunk 5]
```diff
@@ -164,6 +183,12 @@ def calculate_statistics_per_coin(self, closed_trades):
             self.buypoints[trade.pair].append(trade.opened_at)
             self.sellpoints[trade.pair].append(trade.closed_at)
 
+            # Update best/worst trade
+            if trade.profit_ratio > best_worst_trade['best_trade']:
+                best_worst_trade['best_trade'] = trade.profit_ratio
+            if trade.profit_ratio < best_worst_trade['worst_trade']:
+                best_worst_trade['worst_trade'] = trade.profit_ratio
+
```
Comment (geordle) [inline modules/stats/stats.py:191]: Think this is unrelated to the rest, iterating 2 times isn't that bad if it improves readability. I'd move this to a seperate method

File: test/stats/test_coin_insights.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+    # Act
+    stats = fixture.create().analyze()
+
+    # Assert
+    assert math.isclose(stats.coin_res[-1].total_profit_percentage, -38.74375)
+    assert math.isclose(stats.main_results.worst_trade, -38.74375)
+
+def test_best_trade():
+    """Given only winning trades, 'worst trade' should be worst of all trades"""
```
Comment (geordle) [inline test/stats/test_coin_insights.py:50]:
```suggestion
    """Given only winning trades, 'best_trade' should be best of all trades"""
```

File: test/stats/test_coin_insights.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+    # Act
+    stats = fixture.create().analyze()
+
+    # Assert
+    assert math.isclose(stats.coin_res[-1].total_profit_percentage, 71.5175)
+    assert math.isclose(stats.main_results.best_trade, 71.5175)
+
+def test_best_worst_trade_only_wins():
+    """Given only winning trades, 'worst trade' should be worst of all trades"""
```
Comment (geordle) [inline test/stats/test_coin_insights.py:74]: @sheridavandenbent please look at comments

File: test/stats/test_coin_insights.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    fixture = StatsFixture(['COIN/BASE'])
+
+    fixture.frame_with_signals['COIN/BASE'] \
+        .add_entry(open=2, high=2, low=2, close=2, volume=1, buy=1, sell=0) \
+        .add_entry(open=2, high=2, low=1, close=1, volume=1, buy=0, sell=1) \
+        .add_entry(open=1, high=1, low=1, close=1, volume=1, buy=1, sell=0) \
+        .add_entry(open=1, high=6, low=1, close=6, volume=1, buy=0, sell=1) \
+        .add_entry(open=6, high=6, low=6, close=6, volume=1, buy=1, sell=0) \
+        .add_entry(open=6, high=6, low=4, close=4, volume=1, buy=0, sell=1)
+
```
Comment (geordle) [inline test/stats/test_coin_insights.py:155]: This is a bit hard to follow. Maybe simplify case/state a bit so it's easier to follow (FI: just 2 entries, multiply by 2 )

File: test/stats/test_main_results.py [hunk 9]
```diff
@@ -149,8 +149,28 @@ def test_stoploss():
     stats = fixture.create().analyze()
 
     # Assert
-    assert stats.main_results.end_capital == 73.5075
+    assert stats.main_results.end_capital == 74.25
+
+def test_trailing_stoploss():
```
Comment (geordle) [inline test/stats/test_main_results.py:154]: I don't think this fully covers `trailing_stoploss`. I'd add a `buy => Rise => Low`

**Review by geordle — CHANGES_REQUESTED** [review]

File: modules/output/results.py [hunk 10]
```diff
@@ -35,8 +35,8 @@ class MainResults:
...
+    best_trade: float
```
[the remark this answers is not in the document]
Comment (sheridavandenbent) [reply modules/output/results.py]: Has been changed! Closing conversation

File: modules/stats/stats.py (hunk 2 above)
Comment (sheridavandenbent) [reply modules/stats/stats.py]: Done in recent commit! Closing conversation

File: modules/stats/stats.py (hunk 4 above)
Comment (sheridavandenbent) [reply modules/stats/stats.py]: Resolved in recent commit, closing conversation

File: modules/stats/stats.py (hunk 5 above)
Comment (sheridavandenbent) [reply modules/stats/stats.py]: Have used the max function! Committed and pushed. Closing conversation!

File: test/stats/test_coin_insights.py (hunk 7 above)
Comment (sheridavandenbent) [reply test/stats/test_coin_insights.py]: Comments are fixed, closing conversation

File: modules/setup/config/__init__.py [hunk 11]
```diff
@@ -57,9 +57,9 @@ def __init__(self, args):
 
     def load_btc_marketchange(self):
         print("[INFO] Fetching marketchange of BTC/USDT...")
-        begin_data = self.exchange.fetch_ohlcv(symbol='BTC/USDT', timeframe='1m', since=self.backtesting_from, limit=1)
```
[the remark this answers is not in the document]
Comment (sheridavandenbent) [reply modules/setup/config/__init__.py:60]: @noudcorten well done, tagging you here so you'll see this! :)
Resolving conversation to allow merging

**Review by geordle — APPROVED** [review]
